src/otoole/results/result_package.py

```diff
@@ -765,22 +767,38 @@ def capital_recovery_factor(
         param CapitalRecoveryFactor{r in REGION, t in TECHNOLOGY} :=
                 (1 - (1 + DiscountRateIdv[r,t])^(-1))/(1 - (1 + DiscountRateIdv[r,t])^(-(OperationalLife[r,t])));
     """
-    if regions and technologies:
-        index = pd.MultiIndex.from_product(
-            [regions, technologies], names=["REGION", "TECHNOLOGY"]
-        )
+
+    def calc_crf(df, operational_life):
+        rate = df["VALUE"] + 1
+        numerator = 1 - rate.pow(-1)
+        denominator = 1 - rate.pow(-operational_life)
+
+        return numerator / denominator
+
+    if not regions and not technologies:
+        return pd.DataFrame(
+            data=[],
+            columns=["REGION", "TECHNOLOGY", "VALUE"],
+        ).set_index(["REGION", "TECHNOLOGY"])
+
+    index = pd.MultiIndex.from_product(
+        [regions, technologies], names=["REGION", "TECHNOLOGY"]
+    )
+    if "TECHNOLOGY" in discount_rate_idv.index.names:
         crf = discount_rate_idv.reindex(index)
-        crf["RATE"] = crf["VALUE"] + 1
-        crf["NUMER"] = 1 - crf["RATE"].pow(-1)
-        crf["DENOM"] = 1 - crf["RATE"].pow(-operational_life["VALUE"])
-        crf["VALUE"] = (crf["NUMER"] / crf["DENOM"]).round(6)
-        return crf.reset_index()[["REGION", "TECHNOLOGY", "VALUE"]].set_index(
-            ["REGION", "TECHNOLOGY"]
-        )
+        crf["VALUE"] = calc_crf(crf, operational_life["VALUE"])
+
     else:
-        return pd.DataFrame([], columns=["REGION", "TECHNOLOGY", "VALUE"]).set_index(
-            ["REGION", "TECHNOLOGY"]
-        )
+        values = discount_rate_idv["VALUE"].copy()
+        crf = discount_rate_idv.reindex(index)
+        # This is a hack to get around the fact that the discount rate is
+        # indexed by REGION and not REGION, TECHNOLOGY
+        crf[::1] = values
```

Comment:
Not saying to change this, just a question for my reference! :) 

Is `crf[::1] = values` the same as `crf[:] = values`? Is there a reason to use one over the other, or are they functionally the same (ie. just personal preference)?

Reply:
Because the indexes (multi indexes) do not match, pandas will not automatically pad out the values. This assignment using `[::1]` forces the dataframe to be filled using the REGION index, matching that from the `values` variable.  I hadn't tried with `[:]`, I'm not sure that would work either, because pandas will still try to match the entries using the index.

Reply:
I tried with [:] and it worked.